File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/I._Анаграммная_группировка/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    words: list[str] = sys.stdin.readline().rstrip().split()
    hashmap: dict[str, list[int]] = dict()
    for idx in range(n):
        base: str = ''.join(sorted(words[idx]))
        if base not in hashmap:
            hashmap[base] = list()
        hashmap[base].append(idx)
    groups: list[list[int]] = sorted(hashmap.values())
    for row_idx in range(len(groups)):
        if row_idx:
            sys.stdout.write('\n')
        for col_idx in range(len(groups[row_idx])):
            if col_idx:
                sys.stdout.write(' ')
            sys.stdout.write(str(groups[row_idx][col_idx]))


if __name__ == '__main__':
    solution()

File: Yandex/Practicum/Algorightms_and_Data_Structures/4_sprint/I._Анаграммная_группировка/solution.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(NlogN)
    Memory Complexity O(N)
    */
    int n;
    std::cin >> n;
    std::vector<std::string> words(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> words[i];
    }
    std::unordered_map<std::string, std::vector<int>> hashmap;
    for (int idx = 0; idx < n; ++idx) {
        std::sort(words[idx].begin(), words[idx].end());
        hashmap[words[idx]].push_back(idx);
    }
    std::vector<std::vector<int>> groups;
    for (auto& pair : hashmap) {
        groups.push_back(pair.second);
    }
    std::sort(groups.begin(), groups.end());
    for (int rowIdx = 0; rowIdx < groups.size(); ++rowIdx) {
        if (rowIdx) {
            std::cout << "\n";
        }
        for (int colIdx = 0; colIdx < groups[rowIdx].size(); ++colIdx) {
            if (colIdx) {
                std::cout << " ";
            }
            std::cout << groups[rowIdx][colIdx];
        }
    }
}

int main() {
    solution();
}